### ISSUE TITLE
Change from mocha to rspec-mocks

### ISSUE THREAD

**Issue by petems** [voxpupuli/plumbing#78]
**tl;dr** I propose we move Vox Pupuli Puppet modules to use rspec-mocks instead of Mocha. It's not super difficult, and only really affects modules with Ruby specific testing, rather than rspec-puppet. So custom facts, functions or type and provider setups. We can investigate if there is existing tooling to do this automatically, similar to [transpec](https://github.com/yujinakayama/transpec).
## Context

In ye olden Ruby days, Mocha was the mocking framework of choice for Ruby testing. There were a lot of things that people liked Mocha for, one of them being stubing of any instance of a class:

``` ruby
# rspec-mock
stub_car = mock(Car)
stub_car.stub!(:speed).and_return(100)
Car.stub!(:new).and_return(stub_car)

# mocha
Car.any_instance.stubs(:speed).returns(100)
```

Taken from http://stackoverflow.com/questions/1406692/what-are-the-advantages-of-mocha-over-rspecs-built-in-mocking-framework

Since then, rspec-mock has come leaps and bounds, more documented and maintained by the core rspec team. It also has 1:1 feature parity with mocha at this stage it seems.

Mocha has some issues with Ruby 2.3.1 that has caused some travis failures (although it appears to be a bug in Ruby rather than a mocha issue causing these failures, more context: https://github.com/voxpupuli/puppet-nginx/issues/917)

Puppet Engineering is aiming to move completely away from mocha to rspec-mock: https://tickets.puppetlabs.com/browse/MODULES-3958

I propose we move Vox Pupuli Puppet modules to use rspec-mocks instead of Mocha. It's not super difficult, and only really affects modules with Ruby specific testing, rather than rspec-puppet. So custom facts, functions or type and provider setups.

An example of this in action is shown in the nginx module: https://github.com/voxpupuli/puppet-nginx/pull/920

**Comment by bastelfreak** [comment]
+1, we already discussed it yesterday on IRC. I think it is just a huge amount of pain to transfer.

**Comment by petems** [comment]
Yeah, its not super high prority and its a bit of a pain. Hopefully theres some sort of automated way of changing...

**Comment by ekohl** [comment]
By now most module are using rspec-mock or are explicitly configured to use mocha:
```console
$ rg mocha modules/voxpupuli/puppet-*/.sync.yml
modules/voxpupuli/puppet-zabbix/.sync.yml
8:  mock_with: ":mocha"

modules/voxpupuli/puppet-windowsfeature/.sync.yml
4:  mock_with: ":mocha"

modules/voxpupuli/puppet-selinux/.sync.yml
6:  mock_with: ':mocha'

modules/voxpupuli/puppet-rabbitmq/.sync.yml
5:  mock_with: ":mocha"

modules/voxpupuli/puppet-nodejs/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-firewalld/.sync.yml
3:  mock_with: ':mocha'
7:      - gem: 'mocha'

modules/voxpupuli/puppet-filemapper/.sync.yml
3:  mock_with: ":mocha"

modules/voxpupuli/puppet-erlang/.sync.yml
7:  mock_with: ":mocha"

modules/voxpupuli/puppet-corosync/.sync.yml
3:  mock_with: ":mocha"

modules/voxpupuli/puppet-cassandra/.sync.yml
3:  mock_with: ":mocha"
```

**Comment by binford2k** [comment]
Is this ticket still relevant?

**Comment by ekohl** [comment]
The list has grown a little bit, probably because we adopted more modules:
```console
$ rg mocha modules/voxpupuli/puppet-*/.sync.yml
modules/voxpupuli/puppet-zabbix/.sync.yml
8:  mock_with: ':mocha'

modules/voxpupuli/puppet-rabbitmq/.sync.yml
4:  mock_with: ':mocha'

modules/voxpupuli/puppet-redis/.sync.yml
8:  mock_with: ":mocha"

modules/voxpupuli/puppet-firewalld/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-erlang/.sync.yml
5:  mock_with: ':mocha'

modules/voxpupuli/puppet-cassandra/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-corosync/.sync.yml
13:  mock_with: ":mocha"

modules/voxpupuli/puppet-windowsfeature/.sync.yml
4:  mock_with: ':mocha'

modules/voxpupuli/puppet-cvmfs/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-nodejs/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-swap_file/.sync.yml
4:  mock_with: ':mocha'

modules/voxpupuli/puppet-selinux/.sync.yml
3:  mock_with: ':mocha'

modules/voxpupuli/puppet-filemapper/.sync.yml
3:  mock_with: ':mocha'
```

Question is: do we want to push for this? Today [puppetlabs_spec_helper](https://github.com/puppetlabs/puppetlabs_spec_helper) pulls in mocha, but it's deprecated. A next major version could drop it but I've been toying with the thought to drop https://github.com/puppetlabs/puppetlabs_spec_helper in our test stack and integrate it into our other testing gems. That could also be another chance to drop mocha as a dependency.

**Comment by petems** [comment]
Man, blast from the past, I opened this issue originally almost 10 years ago!

I've been playing with LLM Agents for this kinda stuff on my open source repos, with Cursor I've been able to switch between frameworks and generate more test coverage pretty easily, and I have a license to use from our AI DevX team at Datadog right now so happy to use it to move a bunch of stuff from mocha -> rspec-mocks, I'll run a quick example with swap_file since that was one of mine originally.

**Comment by binford2k** [comment]
@petems holy crap, I didn't expect you to pop up! I'd love to get you involved in the community again. We're working on getting datadog integrated into our CI now. (https://github.com/voxpupuli/community-triage/issues/17)

**Comment by TheMeier** [comment]
is this actually needed:
```
spec/spec_helper.rb:
  mock_with: ':rspec'
```
seen it in a few modules

**Comment by TheMeier** [comment]
All modules have been addressed except for [cassandra](https://github.com/voxpupuli/puppet-cassandra) and [nagios_proviers](https://github.com/voxpupuli/puppet-nagios_providers) because the need archiving or reviving.

IMHO this can be closed
